File: Api/Api/Services/CommentService.cs
using Api.Dtos;
using Api.Entities;
using Api.Interfaces.Repository;
using Api.Interfaces.Service;
using Api.Mapper;
using Api.Requests.Comment;
using Microsoft.AspNetCore.Http.HttpResults;

namespace Api.Services
{
    public class CommentService : ICommentService
    {
        ICommentRepository _commentRepo;

        public CommentService(ICommentRepository commentRepo)
        {
            _commentRepo = commentRepo;
        }
        public async Task<List<CommentDto>> GetAllAsync()
        {
            List<Comment> comments = await _commentRepo.GetAllAsync();

            List<CommentDto> commentDtos = comments.Select(c => c.ToCommentDto()).ToList();

            return commentDtos;
        }

        public async Task<List<CommentDto>> GetRecipeCommentsAsync(int recipeId)
        {
            List<Comment> commentModels = await _commentRepo.GetRecipeCommentsAsync(recipeId);

            if (commentModels is null)
            {
                throw new KeyNotFoundException($"No comments found for recipe with ID {recipeId}.");
            }

            List<CommentDto> commentDtos = commentModels.Select(cm => cm.ToCommentDto()).ToList();

            return commentDtos;
        }

        public async Task<CommentDto> CreateAsync(CommentCreateRequest? commentRequest)
        {
            if (commentRequest is null)
            {
                throw new ArgumentNullException(nameof(commentRequest), "Comment data cannot be null.");
            }

            try
            {
                Comment commentModel = commentRequest.ToCommentFromCommentCreateRequest();

                Comment createdCommentModel = await _commentRepo.CreateAsync(commentModel);

                return createdCommentModel.ToCommentDto();
            }
            catch (Exception ex)
            {
                throw new Exception("An error occurred while creating the comment.", ex);
            }
        }

        //public Task<CommentDto> Delete(int id)
        //{
        //    throw new NotImplementedException();
        //}

        
    }
}

File: Client/src/services/commentService.ts
import axios from 'axios';
import { RecipeComment } from '../interfaces/RecipeComment';
import { CommentCreate } from '../interfaces/CommentCreate';

const API_PREFIX = "http://localhost:5098/api/comments";

export const getRecipeComments = async (recipeId: number): Promise<RecipeComment[]> => {
  const comments: RecipeComment[] = await axios.get(`${API_PREFIX}/recipe-comments/${recipeId}`)
    .then(response => {
      return response.data;
    })
    .catch(error => {console.error(error)});

  if (!comments) { throw new Error(`with id ${recipeId} not found.`); }

  return comments;
};

export const addComment = async (body: CommentCreate): Promise<RecipeComment> => {
  const comment: RecipeComment =  await axios.post(`${API_PREFIX}`, body)
    .then(res => {
      return res.data;
    })
    .catch(error => {
      console.error('ERROR: ', error)
    });

  return comment;
};
